src/signature/spki.rs

```diff
@@ -0,0 +1,202 @@
+}
+
+/// Parse a public key in the DER-encoded ASN.1 `SubjectPublicKeyInfo`
+/// format described in [RFC 5280 Section 4.1], which is a sequence of an
+/// `AlgorithmIdentifier` and the key value.
+///
+/// If the `AlgorithmIdentifier` in the SPKI does not match the provided
+/// `signature_alg`, or if the DER encoding is invalid, an error will be
+/// returned.
+///
+/// If the function returns successfully, the `key_value` field in the
+/// resulting `SubjectPublicKeyInfo` struct is suitable for use with
+/// `signature::verify()`.
+///
+/// A common situation where this encoding is encountered is when using public
+/// keys exported by OpenSSL. If you export an RSA or ECDSA public key from a
+/// keypair with `-pubout` and friends, you will get DER-encoded
+/// `SubjectPublicKeyInfo`.
+///
+/// [RFC 5280 Section 4.1]: https://tools.ietf.org/html/rfc5280#section-4.1
+pub fn parse_spki<'a>(signature_alg: &Algorithm, public_key_spki: untrusted::Input<'a>)
+        -> Result<SubjectPublicKeyInfo<'a>, ParseSPKIError> {
+    let unwrapped_spki_der = try!(public_key_spki.read_all(ParseSPKIError::BadDER, |input| {
+        der::expect_tag_and_get_value(input, der::Tag::Sequence)
+            .map_err(|_| ParseSPKIError::BadDER)
+    }));
+
+    let spki = try!(parse_spki_value(unwrapped_spki_der));
+    if !signature_alg.public_key_alg_id
+        .matches_algorithm_id_value(spki.algorithm_id_value) {
+        return Err(ParseSPKIError::UnsupportedSignatureAlgorithmForPublicKey);
+    }
+
+    Ok(spki)
+}
+
+/// Represents the contents of `SubjectPublicKeyInfo` described in
+/// RFC 5280 Section 4.1: https://tools.ietf.org/html/rfc5280#section-4.1
+#[derive(Debug)]
+pub struct SubjectPublicKeyInfo<'a> {
+    /// The algorithm id ASN.1.
+    pub algorithm_id_value: untrusted::Input<'a>,
+    /// The key ASN.1 bit string.
+    pub key_value: untrusted::Input<'a>,
+}
+
+// Parse the public key into an algorithm OID, an optional curve OID, and the
+// key value. The caller needs to check whether these match the
+// `PublicKeyAlgorithm` for the `SignatureAlgorithm` that is matched when
+// parsing the signature.
+fn parse_spki_value(input: untrusted::Input)
+                    -> Result<SubjectPublicKeyInfo, ParseSPKIError> {
+    input.read_all(ParseSPKIError::BadDER, |input| {
+        let algorithm_id_value =
+            try!(der::expect_tag_and_get_value(input, der::Tag::Sequence)
+                .map_err(|_| ParseSPKIError::BadDER));
+        let key_value = try!(der::bit_string_with_no_unused_bits(input)
+            .map_err(|_| ParseSPKIError::BadDER));
+        Ok(SubjectPublicKeyInfo {
+            algorithm_id_value: algorithm_id_value,
+            key_value: key_value,
+        })
+    })
+}
+
+/// Groups an ASN.1 AlgorithmIdentifier and a `ring` VerificationAlgorithm.
+pub struct Algorithm {
+    /// The `algorithm` member in SPKI from https://tools.ietf.org/html/rfc5280#section-4.1.
+    public_key_alg_id: AlgorithmIdentifier,
```

Comment:
This one I didn't make public because I'm not sure why anyone outside this library would need it. But on the other hand, it feels odd to have one private field in a public struct...